docs/decisions/20240103-use-jwt-claims-for-agent-admin-auth.md

```diff
@@ -0,0 +1,155 @@
+- Deciders: Pat Losoponkul, Yurii Shynbuiev, David Poltorak, Shailesh Patil
+- Date: 2024-01-03
+- Tags: multitenancy, authorisation, authentication
+
+Technical Story: [Allow agent admin role to be authenticated by Keycloak JWT | https://input-output.atlassian.net/browse/ATL-6074]
+
+## Context and Problem Statement
+
+External IAM authentication and authorization are currently limited to tenants using the wallets.
+Administrators currently rely on a static API key configured in the agent.
+Enhancing the security of admin access requires extending support for admin authentication through an external IAM.
+This introduces a concept of role to the authentication and authorization process.
+
+The existing tenant authentication and authorization model relies on UMA (user-managed access)
+which defines resources, policies, and permissions.
+Tenants gain access to resources based on defined permissions, effectively controlling resource access.
+In addition to wallet usage, the agent also handles wallet management,
+a functionality utilized by both tenants and administrators.
+While administrators don't directly use the wallet, they oversee its management.
+Integrating an auth model to distinguish between admins and tenants presents a new challenge.
+
+- Where and how to define the role of admin and tenant?
+- What boundary the admin role should be scoped to?
+- How to support different deployment topologies? (e.g. multi-agent in a shared realm, but not share the admin users)
+- How does it interact with the resource scope in the UMA model?
+
+## Decision Drivers
+
+- Must not prevent us from using other IAM systems
+- Must not prevent us from supporting fine-grained tenant wallet access in the future
+- Must allow SSO configuration
+- Should not mix admin access with tenant access
+- Should be easy to setup, configure and maintain
+
+## Considered Options
+
+1. Use `ClientRole` for defining roles in Keycloak
+
+    In this option, the `ClientRole` is configured at the client level,
+    and the user is mapped to the `ClientRole` using a role mapper.
+    The role claim will be available in the JWT token at `resource_access.<client_id>.roles`.
+
+2. Use `RealmRole` for defining roles in Keycloak
+
+    In this option, the `RealmRole` is configured at the realm level,
+    and the user is mapped to the `RealmRole` using a role mapper.
+    The role claim will be available in the JWT token at `realm_acces.roles`.
+
+3. Use custom user attribute for defining roles in Keycloak
+
+    In this option, the role is defined as a user attribute.
+    Then the user attribute will be included in a token using a token mapper at any pre-configured path.
+
+4. Reuse UMA model and define the agent as a resource with the resource scope as a role
+
+    In this option, the agent is defined as another UMA resource and the user role is defined by the agent resource scope.
+
+## Decision Outcome
+
+Option 1: Use `ClientRole` for defining roles in keycloak.
```

Comment:
TL;DR;
ClientRole approach will work as well as RealmRole approach. The only difference between them is the `namespace` isolation inside of the realm that fits well for the sandbox environment (2 agents / 1 realm), and the `role` value path in the JWT document.
I propose implementing both (RealmRole first, ClientRole second) and have the corresponding configuration setup in the Agent.

Reply:
The decision outcome also suggests the role claim path to be configurable. With this, the agent is already agnostic of the Keycloak configuration. RealmRole, ClientRole or Attribute can be supported as long as we know the claim path.